torch/csrc/jit/passes/onnx/shape_type_inference.cpp

```diff
@@ -456,9 +459,133 @@ void ONNXAssignOutputShape(
+// for subblock 1 is inferred from %y.1. Shape and type of Subblock
+// output %7 is inferred from %y.5.
+//
+// graph(%y.1 : Int(3:4, 4:1, requires_grad=0, device=cpu)):
+//   ...
+//   %7 : Tensor = prim::Uninitialized()
+//   %16 : bool, %17 : Tensor, %y.14 : Tensor = prim::If(%15) #
+//   test/onnx/test_pytorch_onnx_onnxruntime.py:614:20
+//     block0():
+//       %y.5 : Tensor = aten::add(%y.1, %3, %6) #
+//       test/onnx/test_pytorch_onnx_onnxruntime.py:615:28
+//       -> (%2, %7, %y.5)
+//     block1():
+//       -> (%1, %y.1, %7)
+//   ...
+
+void ONNXIfShapeTypeInference(Node* node) {
+  for (Block* b : node->blocks()) {
+    for (Node* n : b->nodes()) {
+      if (n->kind() == ::c10::onnx::If) {
+        ONNXIfShapeTypeInference(n);
+      }
+    }
+  }
+
+  if (node->kind() != ::c10::onnx::If) {
+    return;
+  }
+
+  GRAPH_DUMP("Graph before fixing If shape type: ", node->owningGraph());
+  auto* if_node = node;
+  auto* graph = if_node->owningGraph();
+
+  // Check if the input to ONNX If node is node Bool, and insert
+  // cast to Bool if needed.
+  if (!if_node->input()->type()->isSubtypeOf(BoolType::get())) {
+    Node* cast_node = CreateCastToBoolNode(if_node->input(), graph);
+    cast_node->insertBefore(if_node);
+    if_node->replaceInputWith(if_node->input(), cast_node->output());
+  }
+
+  Block* then_block = if_node->blocks()[0];
+  Block* else_block = if_node->blocks()[1];
+
+  // Infer shape and type for subblock outputs
+  TORCH_INTERNAL_ASSERT(
+      then_block->outputs().size() == else_block->outputs().size())
+  for (size_t i = 0; i < else_block->outputs().size(); i++) {
+    Value* then_block_output = then_block->outputs()[i];
+    Value* else_block_output = else_block->outputs()[i];
+
+    // If both subblocks have an uninitialized output, shape and type cannot
+    // be inferred.
+    TORCH_INTERNAL_ASSERT(
+        !(IsUninitializedBlockOutput(then_block_output->node()) &&
+          IsUninitializedBlockOutput(else_block_output->node())),
+        "Cannot infer shape and type for ONNX If with uninitialized output in both subblocks. Please check the model graph.");
+
+    if (IsUninitializedBlockOutput(then_block_output->node())) {
+      InferShapeTypeForUninitializedOutput(
+          graph, then_block, then_block_output, else_block_output);
+      if_node->outputs()[i]->setType(then_block->outputs()[i]->type());
+    } else if (IsUninitializedBlockOutput(else_block_output->node())) {
+      InferShapeTypeForUninitializedOutput(
+          graph, else_block, else_block_output, then_block_output);
+      if_node->outputs()[i]->setType(else_block->outputs()[i]->type());
+    }
+  }
+}
+
 void ONNXShapeTypeInference(std::shared_ptr<Graph>& graph, int opset_version) {
   for (auto n : graph->nodes()) {
     ONNXShapeTypeInference(n, opset_version);
+    if (n->kind() == ::c10::onnx::If || n->kind() == ::c10::onnx::Loop) {
+      ONNXIfShapeTypeInference(n);
```

Comment:
There should not be symbolic shapes dependent here, since this is only handling shape and type for uninitialized node (which has no children).
The reason why we need this as a post pass is that we need to handle nested if/else blocks with the inner if subblock first.